### PR TITLE
kernel timer API rework

### PR DESCRIPTION
**Pull request by rkitover**
@behlendorf this is my first version, want to run it through the CI.

These are the errors I noticed when I was doing the alpine testing, I happened to have 5.0-rc7 in that image.

I tested compiling with 4.20, 4.9 and 5.1-rc5.

I don't have a zfs system right now to test on, but I will try to do some testing on my laptop.

I haven't worked on this codebase before, so if anything is out of place or named wrong please let me know and I will fix it.

#### commit message below

kernel timer API rework

In `config/kernel-timer.m4` check more generally for the new
`timer_setup()` APIs, but keep the old timer callback signature check
because some kernels (notably 4.14) have the new `timer_setup()` API but
use the old callback signature. Also add a check for a `flags` member in
`struct timer_list`, much older kernels do not have it.

Add compatibility shims to `include/spl/sys/timer.h` to allow using the
new timer APIs with the only two caveats being that the callback
argument type must be declared as `spl_timer_list_t` and an explicit
assignment is required to get the timer variable for the `timer_of()`
macro. So the callback would look like this:

```c
__cv_wakeup(spl_timer_list_t t)
{
        struct timer_list *tmr = (struct timer_list *)t;
        struct thing *parent = from_timer(parent, tmr,
                parent_timer_field);
        ... /* do stuff with parent */
```

Make some minor changes to `spl-condvar.c` and `spl-taskq.c` to use the
new timer APIs instead of conditional code.

Also, to fix a problem in 5.0-rc7 which was the impetus for this work,
in `config/kernel-pde-data.m4` check that the `PDE_DATA()` is available
in any form rather than checking for the symbol, as it is sometimes
defined as a function.

Signed-off-by: Rafael Kitover <rkitover@gmail.com>

### Motivation and Context
I did this to fix the errors compiling against 5.0-rc7. Also it is my hope that the timer API changes are nicer than the previous iteration.

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by rkitover** [comment]
Looks like I need to check for both `timer_setup()` **AND** the signature of the timer callback.

**Comment by rkitover** [comment]
I think it's ready now, should be the normal CI results. Will wait for your feedback.

File: include/spl/sys/timer.h [hunk 1]
```diff
@@ -72,4 +72,30 @@ usleep_range(unsigned long min, unsigned long max)
 #define	USEC_TO_TICK(us)		usecs_to_jiffies(us)
 #define	NSEC_TO_TICK(ns)		usecs_to_jiffies(ns / NSEC_PER_USEC)
 
+#ifndef from_timer
+#define	from_timer(var, timer, timer_field) \
+	container_of(timer, typeof(*var), timer_field)
+#endif
+
+
```
Comment (behlendorf) [inline include/spl/sys/timer.h]: nit: extra empty line

File: include/spl/sys/timer.h [hunk 2]
```diff
@@ -72,4 +72,30 @@ usleep_range(unsigned long min, unsigned long max)
...
+#ifdef HAVE_KERNEL_TIMER_FUNCTION_TIMER_LIST
+#define	TIMER_FUNC_ARG	struct timer_list *
+#else
+#define	TIMER_FUNC_ARG	unsigned long
+#endif
```
Comment (behlendorf) [inline include/spl/sys/timer.h]: Rather then define this as a macro, can  we add our own typedef for this?  Or does it need to be a macro for the `timer_setup()` prototype below.  

```c
#if defined(HAVE_KERNEL_TIMER_FUNCTION_TIMER_LIST)
typedef struct timer_list *     spl_timer_list_t;
#else
typedef unsigned long           spl_timer_list_t;
#endif
```

**Comment by rkitover** [comment]
@behlendorf thank you for taking a look, I just made these changes and will see what the CI says.

Also I was wondering, do we have documentation for spl? Should we?

**Comment by rkitover** [comment]
@behlendorf I wanted to ask you also, I have a lovely Project Trident install going, and that runs on CURRENT, where are the freebsd ZoL incorporation efforts happening and can I get involved in that.

**Comment by behlendorf** [comment]
@rkitover can you address the one [cstyle issue reported](
http://build.zfsonlinux.org/builders/Ubuntu%2017.10%20x86_64%20%28STYLE%29/builds/13001/steps/shell_2/logs/stdio), then this looks good to me.  We have one other issue with the 5.0.8 kernel causing the Federa failures above.   I should be able to look in to that next week, unless some beats me to it.  So I'll include this change for that testing.

>  do we have documentation for spl?  Should we

Only in the source code.  The public interfaces are designed to match those from illumos in behavior so you can draw on some of that documentation.  But internally, the implementations are different.  I think improving the documentation would definitely be a good thing.

> I have a lovely Project Trident install going, and that runs on CURRENT, where are the freebsd ZoL incorporation efforts happening and can I get involved in that.

Great timing.  There we a call for testing recently send out to the openzfs-developer list which should reference everything you need to get going.  We'd love the feedback and help!

https://openzfs.topicbox.com/groups/developer/T2b2cb7346ab24f1e/call-for-testing-of-freebsd-zol

**Comment by rkitover** [comment]
@behlendorf thank you for the info, I will try to spend some time with the freebsd port soon.

I couldn't figure out before what that style error meant before, but I got it this time.

**Comment by behlendorf** [comment]
> These are the errors I noticed when I was doing the alpine testing, I happened to have 5.0-rc7 in that image.

@rkitover would you posting exactly what build errors this PR resolved on Alpine since we're not seeing a problem in the CI environments.

@kusumi would you mind providing a second review for this as well.

**Comment by rkitover** [comment]
@behlendorf the errors were specific to 5.0-rc7 rather than alpine, I wouldn't have even bothered but I thought it was a different kernel, and hopefully this is a useful refactor anyway. I will post the exact errors in a bit.

File: config/kernel-timer.m4 [hunk 3]
```diff
@@ -1,32 +1,71 @@
 dnl #
-dnl # 4.15 API change
-dnl # https://lkml.org/lkml/2017/11/25/90
-dnl # Check if timer_list.func get passed a timer_list or an unsigned long
-dnl # (older kernels).  Also sanity check the from_timer() and timer_setup()
-dnl # macros are available as well, since they will be used in the same newer
-dnl # kernels that support the new timer_list.func signature.
```
Comment (kusumi) [inline config/kernel-timer.m4]: It looks to me this shouldn't be removed as the `timer_list.func` compile time test still exists.

File: config/kernel-pde-data.m4 [hunk 4]
```diff
@@ -4,11 +4,11 @@ dnl # PDE is replaced by PDE_DATA
 dnl #
 AC_DEFUN([ZFS_AC_KERNEL_PDE_DATA], [
 	AC_MSG_CHECKING([whether PDE_DATA() is available])
-	ZFS_LINUX_TRY_COMPILE_SYMBOL([
+	ZFS_LINUX_TRY_COMPILE([
```
Comment (kusumi) [inline config/kernel-pde-data.m4]: It looks like `PDE_DATA()` first appeared in 3.10 and has always been an exported function, i.e. `EXPORT_SYMBOL(PDE_DATA).`

To be more specific, it was first added as an inline function in upstream commit 
d9dda78bad879595d8c4220a067fc029d6484a16
and then changed to non-inline exported function in upstream commit
c30480b92cf497aa3b463367a82f1c2fdc5c46e9
both of which first appeared in 3.10 in terms of release kernel, so `PDE_DATA()` has always been a exported function.

File: config/kernel-timer.m4 [hunk 5]
```diff
@@ -1,32 +1,71 @@
...
+		struct timer_list tl;
+		tl.flags = 2;
+	],[
+		AC_MSG_RESULT(yes)
+		AC_DEFINE(HAVE_KERNEL_TIMER_LIST_FLAGS, 1,
+		    [struct timer_list has a flags member])
+	],[
+		AC_MSG_RESULT(no)
+	])
+
```
Comment (kusumi) [inline config/kernel-timer.m4]: This change is from a commit in 4.2 kernel.
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=0eeda71bc30d74f66f8231f45621d5ace3419186

**Comment by rkitover** [comment]
This is very strange, with 5.0-rc7 on gentoo master compiles correctly for me. Now I need to figure out what it is that's specific to the alpine environment or the particular commit I was testing at the time to make it fail. I thought these errors were specific to 5.0-rc7 but apparently not.

**Comment by rkitover** [comment]
I cannot reproduce this on alpine either, unfortunately I deleted that kernel tree and am using the git tags now. I have no idea what happened.

In this case, I will remove the `PDE_DATA` change and address kusumi's points and do this as a refactoring change.

**Comment by rkitover** [comment]
@kusumi apologies for the delay in getting back to this, does this look better?

**Comment by rkitover** [comment]
@behlendorf Hi, can this be merged now? Also any other build related issues you'd like me to look at?

**Comment by behlendorf** [comment]
> Also any other build related issues you'd like me to look at?

Since you're offering it would be great if you could review the [outstanding PRs]( https://github.com/zfsonlinux/zfs/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aopen+Linux+5.2+) for Linux 5.2 compatibility.

**Pull request merged**